### PR TITLE
Use Paper command forwarding API

### PR DESCRIPTION
**Pull request by JRoy**
Uses API added in https://github.com/PaperMC/Paper/commit/9940bcabcc33585815fa30feb63d22f19f87acb4, when available.

### CODE REVIEW

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java [hunk 1]
```diff
@@ -1,21 +1,49 @@
...
 import net.kyori.adventure.text.ComponentLike;
 import org.bukkit.Bukkit;
+import org.bukkit.command.CommandSender;
 import org.bukkit.command.ConsoleCommandSender;
 import org.jetbrains.annotations.NotNull;
 
+import java.util.function.Consumer;
+
 public class PaperCommandSender extends BukkitSenderProvider {
+    private static final boolean SHOULD_MAKE_JMP_HAPPY;
```
Comment (mdcfe) [inline providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java]: Not wrong but `FORWARDING_SENDER_AVAILABLE`

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java [hunk 2]
```diff
@@ -25,57 +53,57 @@ public void sendMessage(ComponentLike message) {
...
 
     @Override
-    public void sendMessage(Identified source, Component message, MessageType type) {
+    public void sendMessage(@NotNull Identified source, @NotNull Component message, @NotNull MessageType type) {
         sendDumbComponent(message);
     }
 
     public void sendDumbComponent(Component message) {
-        sendMessage(Bukkit.getUnsafe().legacyComponentSerializer().serialize(message));
+        sendMessage(PaperComponents.legacySectionSerializer().serialize(message));
```
Comment (mdcfe) [inline providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java:106]: Is `PaperComponents` available on all Paper versions we support with Adventure?

Comment (JRoy) [reply providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java:106]: yeah i checked the 1.16.5 javadocs

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java (hunk 1 above)
Comment (JRoy) [reply providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java]:
```suggestion
    private static final boolean FORWARDING_SENDER_AVAILABLE;
```

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java [hunk 3]
```diff
@@ -1,21 +1,49 @@
...
+
+    static {
+        boolean jmpHappy;
+        try {
+            Bukkit.createCommandSender(component -> {});
+            jmpHappy = true;
+        } catch (NoSuchMethodError ignored) {
+            jmpHappy = false;
+        }
+        SHOULD_MAKE_JMP_HAPPY = jmpHappy;
```
Comment (JRoy) [inline providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java]:
```suggestion
        FORWARDING_SENDER_AVAILABLE = jmpHappy;
```

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java [hunk 4]
```diff
@@ -1,21 +1,49 @@
...
+    }
+
     public PaperCommandSender(ConsoleCommandSender base, MessageHook hook) {
         super(base, hook);
     }
 
+    public static boolean shouldMakeJmpHappy() {
+        return SHOULD_MAKE_JMP_HAPPY;
```
Comment (JRoy) [inline providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java]:
```suggestion
    public static boolean forwardingSenderAvailable() {
        return FORWARDING_SENDER_AVAILABLE;
```

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java [hunk 5]
```diff
@@ -1,21 +1,49 @@
...
+    }
+
+    public static CommandSender createCommandSender(Consumer<String> consumer) {
+        if (!shouldMakeJmpHappy()) {
```
Comment (JRoy) [inline providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java]:
```suggestion
        if (!forwardingSenderAvailable()) {
```

File: EssentialsDiscord/src/main/java/net/essentialsx/discord/util/DiscordCommandSender.java [hunk 6]
```diff
@@ -36,11 +37,21 @@ public DiscordCommandSender(JDADiscordService jda, ConsoleCommandSender sender,
         }, 0, 20);
     }
 
+    private CommandSender getCustomSender(final ConsoleCommandSender consoleSender, final BukkitSenderProvider.MessageHook hook) {
+        if (VersionUtil.isPaper() && VersionUtil.getServerBukkitVersion().isHigherThanOrEqualTo(VersionUtil.v1_16_5_R01)) {
+            if (PaperCommandSender.shouldMakeJmpHappy()) {
```
Comment (JRoy) [inline EssentialsDiscord/src/main/java/net/essentialsx/discord/util/DiscordCommandSender.java]:
```suggestion
            if (PaperCommandSender.forwardingSenderAvailable()) {
```

**Review by jpenilla — APPROVED** [review]

File: providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java [hunk 7]
```diff
@@ -1,21 +1,49 @@
...
 
+    public static boolean forwardingSenderAvailable() {
+        return FORWARDING_SENDER_AVAILABLE;
+    }
+
+    public static CommandSender createCommandSender(Consumer<String> consumer) {
+        if (!forwardingSenderAvailable()) {
+            return null;
+        }
+        return Bukkit.createCommandSender(component -> consumer.accept(PaperComponents.legacySectionSerializer().serialize(component)));
```
Comment (zml2008) [inline providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java:42]: legacy :(

Comment (JRoy) [reply providers/PaperProvider/src/main/java/net/ess3/provider/providers/PaperCommandSender.java:42]: No other option for us atm sadly

**Review by mdcfe — APPROVED** [review]

**Pull request merged**